x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/rules_feature_tour_context.tsx

```diff
@@ -0,0 +1,133 @@
+ * 2.0.
+ */
+
+import React, { createContext, useContext, useEffect, useMemo, FC } from 'react';
+
+import { noop } from 'lodash';
+import {
+  useEuiTour,
+  EuiStatelessTourStep,
+  EuiSpacer,
+  EuiButton,
+  EuiTourStepProps,
+} from '@elastic/eui';
+import { invariant } from '../../../../../../common/utils/invariant';
+import { useKibana } from '../../../../../common/lib/kibana';
+
+import * as i18n from '../translations';
+
+export interface RulesFeatureTourContextType {
+  steps: {
+    inMemoryTableStepProps: EuiTourStepProps;
+    bulkActionsStepProps: EuiTourStepProps;
+  };
+  goToNextStep: () => void;
+  finishTour: () => void;
+}
+
+const STORAGE_KEY = 'securitySolution.rulesFeatureTour';
+const TOUR_POPOVER_WIDTH = 360;
+
+const featuresTourSteps: EuiStatelessTourStep[] = [
+  {
+    step: 1,
+    title: i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP_TITLE,
+    content: <></>,
+    stepsTotal: 2,
+    children: <></>,
+    onFinish: noop,
+    maxWidth: TOUR_POPOVER_WIDTH,
+  },
+  {
+    step: 2,
+    title: i18n.FEATURE_TOUR_BULK_ACTIONS_STEP_TITLE,
+    content: <p>{i18n.FEATURE_TOUR_BULK_ACTIONS_STEP}</p>,
+    stepsTotal: 2,
+    children: <></>,
+    onFinish: noop,
+    anchorPosition: 'rightUp',
+    maxWidth: TOUR_POPOVER_WIDTH,
+  },
+];
+
+const tourConfig = {
+  currentTourStep: 1,
+  isTourActive: true,
+  tourPopoverWidth: TOUR_POPOVER_WIDTH,
+  tourSubtitle: i18n.FEATURE_TOUR_TITLE,
+};
+
+const RulesFeatureTourContext = createContext<RulesFeatureTourContextType | null>(null);
+
+/**
+ * Context for new rules features, displayed in demo tour(euiTour)
+ * It has a common state in useEuiTour, which allows transition from one step to the next, for components within it[context]
+ * It also stores tour's state in localStorage
+ */
+export const RulesFeatureTourContextProvider: FC = ({ children }) => {
+  const { storage } = useKibana().services;
+  const initialStore = useMemo(() => storage.get(STORAGE_KEY) ?? tourConfig, [storage]);
+
+  const [stepProps, actions, reducerState] = useEuiTour(featuresTourSteps, initialStore);
+
+  const finishTour = actions.finishTour;
+  const goToNextStep = actions.incrementStep;
+
+  const inMemoryTableStepProps = useMemo(
+    () => ({
+      ...stepProps[0],
+      content: (
+        <>
+          <p>{i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP}</p>
+          <EuiSpacer />
+          <EuiButton color="primary" onClick={goToNextStep}>
+            {i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP_NEXT}
+          </EuiButton>
+        </>
+      ),
+    }),
+    [stepProps, goToNextStep]
+  );
+
+  useEffect(() => {
+    storage.set(STORAGE_KEY, reducerState);
+  }, [reducerState, storage]);
+
+  const providerValue = useMemo(
+    () => ({
+      steps: {
+        inMemoryTableStepProps,
+        bulkActionsStepProps: stepProps[1],
+      },
+      finishTour,
+      goToNextStep,
+    }),
+    [finishTour, goToNextStep, inMemoryTableStepProps, stepProps]
```

Comment:
Just to double-check, is the `stepProps` array being recreated every time any step in it changes?

Reply:
sorry, what do you mean by `being recreated every time any step in it change`?

Reply:
I mean, when the user navigates forward through the tour (goes from step 1 to step 2), `stepProps[0]` will change its prop that tells if it's active or not to false; `stepProps[1]` will change the same prop to true. When it happens:

- does the tour's hook create new instances of objects representing steps?
- does the tour's hook create a new instance of the `stepProps` array, so we could use it as a dependency here?

Reply:
does the tour's hook create new instances of objects representing steps (stepProps[0] and stepProps[1])?
does the tour's hook create a new instance of the stepProps array, so we could use it as a dependency here?

yes, both are true. Otherwise, when we closed a first step, second step would've not appeared